File: windows/src/RealtimeTranslator.Platform/Subtitles/SubtitleTranscriptStore.cs
using System;
using System.Globalization;
using System.IO;
using System.Text;
using RealtimeTranslator.Core.Subtitles;

namespace RealtimeTranslator.Platform.Subtitles;

public enum SubtitleTranscriptAppendResult
{
    Appended,
    SkippedDuplicate,
    SkippedEmpty,
    Capped,
    Failed,
}

/// <summary>確定字幕ペアをローカルファイルへ追記する。件数に比例するメモリは持たない。</summary>
public sealed class SubtitleTranscriptStore
{
    public const int MaxFileBytes = SubtitleTranscriptLimits.MaxFileBytes;
    public const string SizeLimitBanner = SubtitleTranscriptLimits.SizeLimitBanner;
    public const string WriteFailureBanner = SubtitleTranscriptLimits.WriteFailureBanner;

    private readonly string _filePath;
    private readonly Func<DateTimeOffset> _now;
    private readonly int _maxFileBytes;
    private readonly object _sync = new();

    private string? _lastSource;
    private string? _lastTranslation;
    private bool _announcedSizeLimit;
    private long? _cachedByteCount;

    public SubtitleTranscriptStore(
        string? filePath = null,
        Func<DateTimeOffset>? now = null,
        int maxFileBytes = SubtitleTranscriptLimits.MaxFileBytes)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(maxFileBytes);
        _filePath = filePath ?? DefaultFilePath();
        _now = now ?? (() => DateTimeOffset.Now);
        _maxFileBytes = maxFileBytes;
    }

    public string FilePath => _filePath;

    public bool HasEntries
    {
        get
        {
            lock (_sync)
            {
                return FileByteCountLocked() > 0;
            }
        }
    }

    public static string DefaultFilePath()
    {
        var directory = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "RealtimeTranslator",
            "transcripts");
        Directory.CreateDirectory(directory);
        return Path.Combine(directory, "session.txt");
    }

    public static string DefaultExportFileName(DateTimeOffset? now = null)
    {
        var timestamp = now ?? DateTimeOffset.Now;
        return string.Create(
            CultureInfo.InvariantCulture,
            $"subtitles-{timestamp:yyyyMMdd-HHmmss}.txt");
    }

    public SubtitleTranscriptAppendResult MarkSessionStart()
    {
        lock (_sync)
        {
            // 新セッションでは直前セッション末尾との連続重複判定を切る。
            _lastSource = null;
            _lastTranslation = null;
            var timestamp = SubtitleTranscriptFormatter.FormatTimestamp(_now());
            var chunk = SubtitleTranscriptFormatter.FormatSessionStart(timestamp);
            return AppendChunkLocked(chunk, updateLastPair: false, source: null, translation: null);
        }
    }

    public SubtitleTranscriptAppendResult AppendEntry(string sourceText, string translatedText)
    {
        ArgumentNullException.ThrowIfNull(sourceText);
        ArgumentNullException.ThrowIfNull(translatedText);

        lock (_sync)
        {
            var source = sourceText.Trim();
            var translation = translatedText.Trim();
            if (source.Length == 0 || translation.Length == 0)
            {
                return SubtitleTranscriptAppendResult.SkippedEmpty;
            }

            if (sourceText == _lastSource && translatedText == _lastTranslation)
            {
                return SubtitleTranscriptAppendResult.SkippedDuplicate;
            }

            var timestamp = SubtitleTranscriptFormatter.FormatTimestamp(_now());
            // 記録は表示クリップ前の原文・訳文そのものを残す。
            var chunk = SubtitleTranscriptFormatter.FormatEntry(timestamp, sourceText, translatedText);
            return AppendChunkLocked(chunk, updateLastPair: true, sourceText, translatedText);
        }
    }

    public void ExportCopy(string destinationPath)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(destinationPath);

        lock (_sync)
        {
            var directory = Path.GetDirectoryName(destinationPath);
            if (!string.IsNullOrEmpty(directory))
            {
                Directory.CreateDirectory(directory);
            }

            if (!File.Exists(_filePath))
            {
                File.WriteAllText(destinationPath, string.Empty, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
                return;
            }

            File.Copy(_filePath, destinationPath, overwrite: true);
        }
    }

    public void Clear()
    {
        lock (_sync)
        {
            var directory = Path.GetDirectoryName(_filePath);
            if (!string.IsNullOrEmpty(directory))
            {
                Directory.CreateDirectory(directory);
            }

            File.WriteAllText(_filePath, string.Empty, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
            _lastSource = null;
            _lastTranslation = null;
            _announcedSizeLimit = false;
            _cachedByteCount = 0;
        }
    }

    private SubtitleTranscriptAppendResult AppendChunkLocked(
        string chunk,
        bool updateLastPair,
        string? source,
        string? translation)
    {
        var chunkBytes = Encoding.UTF8.GetByteCount(chunk);
        var currentBytes = FileByteCountLocked();
        if (currentBytes >= _maxFileBytes || currentBytes + chunkBytes > _maxFileBytes)
        {
            // macOS 実装と同じく初回/以降を区別してフラグを立てる（呼び出し側の一度だけ案内と併用）。
            if (_announcedSizeLimit)
            {
                return SubtitleTranscriptAppendResult.Capped;
            }

            _announcedSizeLimit = true;
            return SubtitleTranscriptAppendResult.Capped;
        }

        try
        {
            EnsureFileExistsLocked();
            using var stream = new FileStream(
                _filePath,
                FileMode.Append,
                FileAccess.Write,
                FileShare.Read);
            var payload = Encoding.UTF8.GetBytes(chunk);
            stream.Write(payload, 0, payload.Length);
            stream.Flush(flushToDisk: true);
            _cachedByteCount = currentBytes + chunkBytes;
            if (updateLastPair)
            {
                _lastSource = source;
                _lastTranslation = translation;
            }

            return SubtitleTranscriptAppendResult.Appended;
        }
        catch (Exception error) when (error is IOException or UnauthorizedAccessException)
        {
            // 失敗したペアも記憶し、同一再試行で失敗通知を連発しない。
            if (updateLastPair)
            {
                _lastSource = source;
                _lastTranslation = translation;
            }

            return SubtitleTranscriptAppendResult.Failed;
        }
    }

    private void EnsureFileExistsLocked()
    {
        var directory = Path.GetDirectoryName(_filePath);
        if (!string.IsNullOrEmpty(directory))
        {
            Directory.CreateDirectory(directory);
        }

        if (!File.Exists(_filePath))
        {
            File.WriteAllText(_filePath, string.Empty, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
            _cachedByteCount = 0;
        }
    }

    private long FileByteCountLocked()
    {
        if (_cachedByteCount is { } cached)
        {
            return cached;
        }

        if (!File.Exists(_filePath))
        {
            _cachedByteCount = 0;
            return 0;
        }

        var size = new FileInfo(_filePath).Length;
        _cachedByteCount = size;
        return size;
    }
}

File: RealtimeTranslator/Subtitles/SubtitleTranscriptStore.swift
import Foundation

enum SubtitleTranscriptAppendResult: Equatable, Sendable {
    case appended
    case skippedDuplicate
    case skippedEmpty
    case capped
    case failed
}

/// 確定字幕ペアをローカルファイルへ追記する。件数に比例するメモリは持たない。
final class SubtitleTranscriptStore: @unchecked Sendable {
    static var sizeLimitBanner: String { SubtitleTranscriptLimits.sizeLimitBanner }
    static var writeFailureBanner: String { SubtitleTranscriptLimits.writeFailureBanner }
    static var maxFileBytes: Int { SubtitleTranscriptLimits.maxFileBytes }

    private let fileURL: URL
    private let now: () -> Date
    private let timeZone: TimeZone
    private let maxFileBytes: Int
    private let lock = NSLock()

    private var lastSource: String?
    private var lastTranslation: String?
    private var announcedSizeLimit = false
    private var cachedByteCount: Int?

    init(
        fileURL: URL,
        now: @escaping () -> Date = { Date() },
        timeZone: TimeZone = .current,
        maxFileBytes: Int = SubtitleTranscriptLimits.maxFileBytes
    ) {
        self.fileURL = fileURL
        self.now = now
        self.timeZone = timeZone
        self.maxFileBytes = maxFileBytes
    }

    var hasEntries: Bool {
        lock.lock()
        defer { lock.unlock() }
        return fileByteCountLocked() > 0
    }

    /// Application Support 下の既定パス。
    static func defaultFileURL(
        fileManager: FileManager = .default,
        bundleIdentifier: String? = Bundle.main.bundleIdentifier
    ) throws -> URL {
        guard let bundleIdentifier else {
            throw CocoaError(.fileNoSuchFile)
        }
        guard let applicationSupport = fileManager.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        ).first else {
            throw CocoaError(.fileNoSuchFile)
        }
        let directory = applicationSupport
            .appendingPathComponent(bundleIdentifier, isDirectory: true)
            .appendingPathComponent("transcripts", isDirectory: true)
        try fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory.appendingPathComponent("session.txt", isDirectory: false)
    }

    @discardableResult
    func markSessionStart() -> SubtitleTranscriptAppendResult {
        lock.lock()
        defer { lock.unlock() }
        // 新セッションでは直前セッション末尾との連続重複判定を切る。
        lastSource = nil
        lastTranslation = nil
        let timestamp = SubtitleTranscriptFormatter.formatTimestamp(now(), timeZone: timeZone)
        let chunk = SubtitleTranscriptFormatter.formatSessionStart(timestamp: timestamp)
        return appendChunkLocked(chunk, updatingLastPair: nil)
    }

    @discardableResult
    func appendEntry(sourceText: String, translatedText: String) -> SubtitleTranscriptAppendResult {
        lock.lock()
        defer { lock.unlock() }

        let trimmedSource = sourceText.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedTranslation = translatedText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedSource.isEmpty, !trimmedTranslation.isEmpty else {
            return .skippedEmpty
        }
        if sourceText == lastSource, translatedText == lastTranslation {
            return .skippedDuplicate
        }

        let timestamp = SubtitleTranscriptFormatter.formatTimestamp(now(), timeZone: timeZone)
        let chunk = SubtitleTranscriptFormatter.formatEntry(
            timestamp: timestamp,
            sourceText: sourceText,
            translatedText: translatedText
        )
        return appendChunkLocked(chunk, updatingLastPair: (sourceText, translatedText))
    }

    func exportCopy(to destination: URL) throws {
        lock.lock()
        defer { lock.unlock() }
        // 自己コピーは記録を消さないよう no-op にする。
        guard destination.standardizedFileURL != fileURL.standardizedFileURL else {
            return
        }
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: fileURL.path), fileByteCountLocked() > 0 else {
            throw CocoaError(.fileReadNoSuchFile)
        }
        if fileManager.fileExists(atPath: destination.path) {
            try fileManager.removeItem(at: destination)
        }
        try fileManager.copyItem(at: fileURL, to: destination)
    }

    func clear() throws {
        lock.lock()
        defer { lock.unlock() }
        let fileManager = FileManager.default
        let directory = fileURL.deletingLastPathComponent()
        try fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
        try Data().write(to: fileURL, options: .atomic)
        lastSource = nil
        lastTranslation = nil
        announcedSizeLimit = false
        cachedByteCount = 0
    }

    /// 既定の書き出しファイル名 `subtitles-YYYYMMDD-HHmmss.txt`。
    static func defaultExportFileName(now: Date = Date(), timeZone: TimeZone = .current) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = timeZone
        formatter.dateFormat = "yyyyMMdd-HHmmss"
        return "subtitles-\(formatter.string(from: now)).txt"
    }

    private func appendChunkLocked(
        _ chunk: String,
        updatingLastPair: (String, String)?
    ) -> SubtitleTranscriptAppendResult {
        let chunkData = Data(chunk.utf8)
        let currentBytes = fileByteCountLocked()
        if currentBytes >= maxFileBytes
            || currentBytes + chunkData.count > maxFileBytes
        {
            if announcedSizeLimit {
                return .capped
            }
            announcedSizeLimit = true
            return .capped
        }

        do {
            try ensureFileExistsLocked()
            let handle = try FileHandle(forWritingTo: fileURL)
            defer { try? handle.close() }
            try handle.seekToEnd()
            try handle.write(contentsOf: chunkData)
            try handle.synchronize()
            cachedByteCount = currentBytes + chunkData.count
            if let updatingLastPair {
                lastSource = updatingLastPair.0
                lastTranslation = updatingLastPair.1
            }
            return .appended
        } catch {
            // 失敗したペアも記憶し、ticker からの同一再試行で失敗バナーを連発しない。
            if let updatingLastPair {
                lastSource = updatingLastPair.0
                lastTranslation = updatingLastPair.1
            }
            return .failed
        }
    }

    private func ensureFileExistsLocked() throws {
        let fileManager = FileManager.default
        let directory = fileURL.deletingLastPathComponent()
        try fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
        if !fileManager.fileExists(atPath: fileURL.path) {
            try Data().write(to: fileURL, options: .atomic)
            cachedByteCount = 0
        }
    }

    private func fileByteCountLocked() -> Int {
        if let cachedByteCount {
            return cachedByteCount
        }
        let values = try? fileURL.resourceValues(forKeys: [.fileSizeKey])
        let size = values?.fileSize ?? 0
        cachedByteCount = size
        return size
    }
}
